### ISSUE TITLE
Middlware is not compatible with fastify reply

### ISSUE THREAD

**Issue by StarpTech** [fastify/fastify#599]
Look at issue https://github.com/fastify/fastify/issues/598

- ~~`reply` interface does not respect the raw `response` state. It also tries to change the response object even when the request was already finished inside a middlware e.g we can't calling `.writeHead()` or `.end()` inside a middleware.~~
- `reply` interface [overrides](https://github.com/fastify/fastify/blob/master/lib/reply.js#L159) any status code smaller than 400 is interpreted as an 500 when an error is set

**Update:** Due to wrong use of middleware, I step into some pitfalls. Only one issue is left.

**Comment by mcollina** [comment]
I’m a bit lost. Can you specify one or more examples that are broken right now?

**Comment by StarpTech** [comment]
**Example 1**: When a route could not be found we can not change the status code inside a middleware because the `reply` does not respect it.
**Reason**: The default notfound handler set the status code to 404 and reply with an error. In `reply` implementation the status code from the raw `response` is not accepted when an error was set [code](https://github.com/fastify/fastify/blob/master/lib/reply.js#L159)

```js
fastify.use((req, res, next) => {
  if (req.url.substr(-1) == '/' && req.url.length > 1) {
      res.setHeader('Location', 'http://127.0.0.1' + fastify.server.address().port + req.url.slice(0, -1))
      res.statusCode = 302
  }
  next();
})
```

**Example 2**: We can not call `.writeHead()` or `res.end()` because the `reply` interface does not check whether the response was already sent.
**Reason**: In the `reply` implementation we don't check if the response was already finished.

```js
fastify.use((req, res, next) => {
  if (req.url.substr(-1) == '/' && req.url.length > 1) {
      res.writeHead(302, {'Location': 'http://127.0.0.1' + fastify.server.address().port + req.url.slice(0, -1)});
  }
  next();
})
```
Will lead to `Error: Can't set headers after they are sent.`

**Comment by nwoltman** [comment]
> **Example 2**: We can not call `.writeHead()` or `res.end()`

Middleware and hooks shouldn't call `res.write()`, `res.writeHead()`, or `res.end()` and continue with the lifecycle because that breaks Fastify's lifecycle expectations (handlers and `onSend` hooks expect to be able to set headers and set/change the payload). If a user does want to use those methods, they should call `res.end()` and **not** `next()` to escape from Fastify's lifecycle (which is what most Express middleware that send a response do).

**Comment by StarpTech** [comment]
Correct If I'm wrong but the whole idea behind supporting standard middlewares like in express is that you can do everything what you want doesn't matter which framework you will use.

> Middleware functions can perform the following tasks: Execute any code.
Make changes to the request and the response objects.
End the request-response cycle.
Call the next middleware in the stack.
If the current middleware function does not end the request-response cycle, it must call next() to pass control to the next middleware function. Otherwise, the request will be left hanging.

Source: https://expressjs.com/en/guide/writing-middleware.html

**Comment by StarpTech** [comment]
Is it possible to call `.end()` inside middleware without calling `.next()`? I think this will lead to an error too.

**Comment by nwoltman** [comment]
> Is it possible to call `.end()` inside middleware without calling `.next()`?

Yes, this is perfectly okay. This is how middleware like [`serve-favicon`](https://github.com/expressjs/serve-favicon) work (I used to use this middleware with Fastify and it works just fine).

**Comment by nwoltman** [comment]
> **Example 1**: When a route could not be found we can not change the status code inside a middleware

I think the correct solution would be to set a custom 404 handler with `fastify.setNotFoundHandler()` to handle the request instead of using a middleware.

Example:

```js
fastify.setNotFoundHandler(function ({req}, reply) {
  if (req.url.substr(-1) == '/' && req.url.length > 1) {
    reply.redirect(req.url.slice(0, -1))
    return
  }
  reply.code(404).send('Not found')
})
```

**Comment by StarpTech** [comment]
When it works then it was an issue by myself. We should also able to write to the response (write, writeHeas) . I can't test it right now.

I think ignoring status codes smaller than 400 when an error is set is also wrong this place should be refactored.

**Comment by StarpTech** [comment]
@nwoltman your workaround is a great but it would be more powerful when we don't need such special handlers because it's not obvious if you work with middlewares.

**Comment by StarpTech** [comment]
@nwoltman I misuse the middleware :S That's all valid examples.

```
fastify.use((req, res, next) => {
  if (req.url.substr(-1) == '/' && req.url.length > 1) {
      res.write("test");
      res.end()
      return
  }
  next();
})
fastify.use((req, res, next) => {
  if (req.url.substr(-1) == '/' && req.url.length > 1) {
      res.writeHead(300);
      res.end()
      return
  }
  next();
})
```

**Comment by mcollina** [comment]
@StarpTech I'm still lost. What is happening in that example that you expect to happen?

**Comment by StarpTech** [comment]
This issue can be closed. I misuse the middleware that's the problem. The clean up is implemented in #601
